cpp/src/parquet/metadata.cc

```diff
@@ -793,26 +793,29 @@ class FileMetaData::FileMetaDataImpl {
 
     std::shared_ptr<FileMetaData> out(new FileMetaData());
     out->impl_ = std::make_unique<FileMetaDataImpl>();
-    out->impl_->metadata_ = std::make_unique<format::FileMetaData>();
-
-    auto metadata = out->impl_->metadata_.get();
-    metadata->version = metadata_->version;
-    metadata->schema = metadata_->schema;
+    {
+      // GH-39336: Copy the metadata, but only the row groups we need.
+      auto temp_row_groups = std::move(this->metadata_->row_groups);
```

Comment:
What is the point of doing this? Overall, are these changes actually bringing a tangible benefit?

Reply:
My problem is that you're temporarily mutating `this` in a method that's not supposed to mutate it. This is brittle (look at that try/except) and also thread-unsafe.

Reply:
You're right. Maybe keeping the current code is ok